banyan/src/forest/read.rs

```diff
@@ -162,44 +183,39 @@ where
                         self.query
                             .intersecting(start_offset, index, &mut q_matching);
                         debug_assert_eq!(branch.children.len(), q_matching.len());
-                        let _ = std::mem::replace(matching, q_matching);
+                        head.filter.replace(q_matching);
 
                         if matches!(self.mode, Mode::Backward) {
-                            *pos = branch.children.len();
+                            head.position = branch.children.len();
                         }
                     }
 
                     let next_idx = match self.mode {
-                        Mode::Forward => *pos,
-                        Mode::Backward => *pos - 1,
+                        Mode::Forward => head.position,
+                        Mode::Backward => head.position - 1,
                     };
-                    if matching[next_idx] {
+                    let should_descend = head.filter.as_ref().expect("not empty")[next_idx];
```

Comment:
maybe instead of wrapping filter in an option, have it as a non-option and have a separate bool to track if it is initialized?

Not sure, just an idea.

Reply:
Hm, but that's basically just putting the bit from the Option to an explicit bit. The smallvec will be only allocated once in any case. I don't see the benefit.

Reply:
yes, I was just thinking that it might look nicer. It compiles to pretty much the same thing.